flow/display_list_benchmarks.cc

```diff
@@ -0,0 +1,1000 @@
+                     SkPoint::Make(length - i % length, length));
+  }
+
+  auto display_list = builder.Build();
+
+  // We only want to time the actual rasterization.
+  for (auto _ : state) {
+    display_list->RenderTo(canvas);
+    canvas_provider->GetSurface()->flushAndSubmit(true);
+  }
+
+  auto filename = canvas_provider->BackendName() + "-DrawLine-" +
+                  std::to_string(state.range(0)) + ".png";
+  canvas_provider->Snapshot(filename);
+}
+
+// Draws a series of shrinking square rects from the requested width down
+// to zero width and repeats until `kRectsToDraw` rects have been drawn.
+// Half the drawn rects will not have an integral offset.
+//
+// The resulting image will be a fully shaded canvas.
+void BM_DrawRect(benchmark::State& state,
+                 std::unique_ptr<CanvasProvider> canvas_provider) {
+  DisplayListBuilder builder;
+  size_t length = state.range(0);
+  canvas_provider->InitializeSurface(length, length);
+  auto canvas = canvas_provider->GetSurface()->getCanvas();
+
+  // The rects will shrink by the inset constant for each iteration. The
+  // inset constant is 0.5 to test rasterisation in between the pixel grids.
+  const SkScalar inset = 0.5;
+  SkRect rect = SkRect::MakeLTRB(0, 0, length, length);
+
+  for (size_t i = 0; i < kRectsToDraw; i++) {
+    // As rects have SkScalar dimensions, we want to ensure that we also
+    // draw rects with non-integer position and size
+    builder.drawRect(rect);
+    rect.inset(inset, inset);
+    if (rect.width() <= 5.0f) {
+      rect = SkRect::MakeLTRB(0, 0, length, length);
+    }
+  }
+
+  auto display_list = builder.Build();
+
+  // We only want to time the actual rasterization.
+  for (auto _ : state) {
+    display_list->RenderTo(canvas);
+    canvas_provider->GetSurface()->flushAndSubmit(true);
+  }
+
+  auto filename = canvas_provider->BackendName() + "-DrawRect-" +
+                  std::to_string(state.range(0)) + ".png";
+  canvas_provider->Snapshot(filename);
+}
+
+// Draws a series of shrinking ovals from the requested width down
+// to zero width and repeats until `kOvalsToDraw` ovals have been drawn.
+// Half the drawn ovals will not have an integral offset.
+//
+// The resulting image will be a fully shaded oval with the bounding box of
+// the entire canvas.
+void BM_DrawOval(benchmark::State& state,
+                 std::unique_ptr<CanvasProvider> canvas_provider) {
+  DisplayListBuilder builder;
+  size_t length = state.range(0);
+  canvas_provider->InitializeSurface(length, length);
+  auto canvas = canvas_provider->GetSurface()->getCanvas();
+
+  SkRect rect;
+  SkScalar scalar_length = length;
+
+  // The first drawOval call draws them shrinking in height, and the second
+  // drawOval call draws them shrinking in width.
+  //
+  // We don't want to shrink them both at the same time because that would just
+  // degenerate to drawCircle.
+  for (size_t i = 0; i < kOvalsToDraw; i++) {
```

Comment:
oh, I forgot to push my latest commits. This code is all redone.

Reply:
I'm not sure what you mean by this? I removed the code where we draw two ovals per loop iteration and now we only draw one. I kept the loop condition at 1000 because I think 2000 is probably too many anyway.

Reply:
I misread. Sorry.